smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaRecordStream.java

```diff
@@ -45,50 +47,49 @@ public void subscribe(
         private final MultiSubscriber<? super ConsumerRecord<K, V>> downstream;
         private final boolean pauseResumeEnabled;
 
-        private final AtomicInteger wip = new AtomicInteger();
         /**
-         * Stores the current downstream demands.
+         * Current state: new (no request yet), polling, paused, cancelled
          */
-        private final AtomicLong requested = new AtomicLong();
-
-        /**
-         * Started flag.
-         */
-        private final AtomicBoolean started = new AtomicBoolean();
+        private final AtomicInteger state = new AtomicInteger(STATE_NEW);
 
+        private final AtomicInteger wip = new AtomicInteger();
         /**
-         * Paused / Resumed flag.
+         * Stores the current downstream demands.
          */
-        private final AtomicBoolean paused = new AtomicBoolean();
+        private final AtomicLong requested = new AtomicLong();
 
         /**
          * The polling uni to avoid re-assembling a Uni everytime.
          */
         private final Uni<ConsumerRecords<K, V>> pollUni;
 
+        private final int maxQueueSize;
+        private final int halfMaxQueueSize;
         private final RecordQueue<ConsumerRecord<K, V>> queue;
         private final long retries;
 
-        /**
-         * {@code true} if the subscription has been cancelled.
-         */
-        private volatile boolean cancelled;
-
         public KafkaRecordStreamSubscription(
                 ReactiveKafkaConsumer<K, V> client,
                 KafkaConnectorIncomingConfiguration config,
                 MultiSubscriber<? super ConsumerRecord<K, V>> subscriber) {
             this.client = client;
             this.pauseResumeEnabled = config.getPauseIfNoRequests();
             this.downstream = subscriber;
+            // Kafka also defaults to 500, but doesn't have a constant for it
             int batchSize = config.config().getOptionalValue(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, Integer.class).orElse(500);
-            this.queue = new RecordQueue<>(2 * batchSize);
+            this.maxQueueSize = batchSize * 2;
+            this.halfMaxQueueSize = batchSize;
+            // we can exceed maxQueueSize by at most 1 batchSize
+            this.queue = new RecordQueue<>(3 * batchSize);
```

Comment:
@ozangunalp I wonder why you allocate 3*batchSize ? It's not max capacity so it could be a little too much, as it should normally not go beyond 2*batchSize, so we may loose some memory due to pre-allocated array.

As the queue is never recreated, if it goes beyond 2*batchSize it will be enlarged once to 50% more capacity (I check on the ArrayDeque code) so it will be the same, but lazily done if needed.

Reply:
If the queue has size `2 * batchSize - 10`, we can add more items to it. Items come in batches, of size at most `batchSize`. If we obtained a batch of max size, the queue would suddenly have `3 * batchSize - 10` items -- and only at that point we would pause.

I thought the comment explains it succinctly but clearly, which obviously isn't the case :-)

Just noting that this sizing algorithm is simple, but not the only one possible. Feel free to suggest a better one.

Reply:
Oh, yes, I now remember that queue.size() is compared to the maxQueueSize to know if pause should be done or not, so we need some overflow capacity anyway as everything happens async, and in batch.